skbio/diversity/beta/_fast_unifrac.py

```diff
@@ -0,0 +1,444 @@
+            return 0.0
+
+    envs = _skbio_counts_to_envs(otu_ids, u_counts, v_counts)
+
+    return fast_unifrac(tree, envs)
+
+
+def weighted_unifrac(u_counts, v_counts, otu_ids, tree, normalized=False,
+                     suppress_validation=False):
+    u_sum = sum(u_counts)
+    v_sum = sum(v_counts)
+
+    if not u_sum or not v_sum:
+        if u_sum + v_sum:
+            # u or v counts are all zeros
+            if normalized:
+                return 1.0
+            # NOTE: we cannot handle the unnormalized case here yet as it
+            # requires operations on the tree vector
+        else:
+            # u and v are zero
+            return 0.0
+    if not suppress_validation:
+        _validate(u_counts, v_counts, otu_ids, tree)
+    envs = _skbio_counts_to_envs(otu_ids, u_counts, v_counts)
+
+    return fast_unifrac(tree, envs, weighted=True, normalized=normalized)
+
+
+def fast_unifrac(t, envs, weighted=False, metric=unifrac, normalized=False):
+    # weighted_unifrac_f=_weighted_unifrac,make_subtree=True):
+    """
+    Run fast unifrac.
+
+    Parameters
+    ----------
+    t: skbio.TreeNode
+        phylogenetic tree relating the sequences.
+    envs: dict
+        dict of {sequence:{env:count}} showing environmental abundance.
+    weighted: bool
+        if True, performs the weighted UniFrac procedure.
+    metric: function
+        distance metric to use.  currently you must use unifrac only
+        if weighted=True.
+        see fast_tree.py for metrics (e.g.: G, unnormalized_G, unifrac, etc.)
+
+    Returns
+    -------
+    u : np.array
+        Unifrac distance matrix
+    """
+    (envs, count_array,
+     unique_envs, env_to_index,
+     node_to_index, env_names,
+     branch_lengths, nodes, t) = _fast_unifrac_setup(t, envs)
+
+    bound_indices = bind_to_array(nodes, count_array)
+    # weighted unifrac
+    if weighted:
+        tip_indices = [n._leaf_index for n in t.tips()]
+        sum_descendants(bound_indices)
+        tip_ds = branch_lengths.copy()[:, np.newaxis]
+        bindings = bind_to_parent_array(t, tip_ds)
+        tip_distances(tip_ds, bindings, tip_indices)
+
+        # NOTE: in the unnormalized case, it is possible for count_array to
+        # have a single column of v_counts come in as all zeros.
+        u_counts = count_array[:, 0]
+        if count_array.shape[1] == 1:
+            v_counts = np.zeros(count_array.shape[0])
+        else:
+            v_counts = count_array[:, 1]
+        u_sum = (np.take(u_counts, tip_indices)).sum()
+        v_sum = (np.take(v_counts, tip_indices)).sum()
+
+        u = w_unifrac(branch_lengths, tip_indices, u_counts, v_counts)
+        if normalized:
+            u /= _branch_correct(tip_ds, u_counts, v_counts, u_sum, v_sum)
+    # unweighted unifrac
+    else:
+        bool_descendants(bound_indices)
+        u = unifrac(branch_lengths, count_array[:, 0], count_array[:, 1])
+
+    return u
+
+
+def _fast_unifrac_setup(t, envs, make_subtree=True):
```

Comment:
Fast UniFrac is only faster when you're computing multiple pairwise distances, right? If that's the case, I wonder if we could somehow hide this stuff under ``pw_distances`` since that's the function that users should be calling to compute UniFrac for more than just a single pair of samples.